allennlp/modules/bimpm_matching.py

```diff
@@ -0,0 +1,396 @@
+    Multi-Perspective Matching for Natural Language Sentences <https://arxiv.org/abs/1702.03814>`_
+    by Zhiguo Wang et al., 2017.
+    Also please refer to the `TensorFlow implementation <https://github.com/zhiguowang/BiMPM/>`_ and
+    `PyTorch implementation <https://github.com/galsang/BIMPM-pytorch>`_.
+
+    Parameters
+    ----------
+    is_forward : ``bool``, required
+        Whether the matching is for forward sequence or backward sequence
+    hidden_dim : ``int``, optional (default = 100)
+        The hidden dimension of the representations
+    num_perspective : ``int``, optional (default = 20)
+        The number of perspective for matching
+    share_weight_between_directions : ``bool``, optional (default = True)
+        If True, share weight between premise to hypothesis and hypothesisto premise,
+        useful for non-symmetric tasks
+    wo_full_match : ``bool``, optional (default = False)
+        If True, exclude full match
+    wo_maxpool_match : ``bool``, optional (default = False)
+        If True, exclude max_pool match
+    wo_attentive_match : ``bool``, optional (default = False)
+        If True, exclude attentive match
+    wo_max_attentive_match : ``bool``, optional (default = False)
+        If True, exclude max attentive match
+    """
+    def __init__(self,
+                 is_forward: bool,
+                 hidden_dim: int = 100,
+                 num_perspective: int = 20,
+                 share_weight_between_directions: bool = True,
+                 wo_full_match: bool = False,
+                 wo_maxpool_match: bool = False,
+                 wo_attentive_match: bool = False,
+                 wo_max_attentive_match: bool = False) -> None:
+        super(BiMPMMatching, self).__init__()
+
+        self.is_forward = is_forward
+        self.hidden_dim = hidden_dim
+        self.num_perspective = num_perspective
+        self.mv_idx_increment = int(not share_weight_between_directions)
+
+        self.wo_full_match = wo_full_match
+        self.wo_maxpool_match = wo_maxpool_match
+        self.wo_attentive_match = wo_attentive_match
+        self.wo_max_attentive_match = wo_max_attentive_match
+
+        if share_weight_between_directions:
+            num_matching = int(not wo_full_match) + \
+                           int(not wo_maxpool_match) + \
+                           int(not wo_attentive_match) + \
+                           int(not wo_max_attentive_match)
+        else:
+            num_matching = int(not wo_full_match) * 2 + \
+                           int(not wo_maxpool_match) + \
+                           int(not wo_attentive_match) * 2 + \
+                           int(not wo_max_attentive_match) * 2
+
+        if num_matching <= 0:
+            raise ConfigurationError("At least one of the matching method should be enabled")
+
+        params = [nn.Parameter(torch.rand(num_perspective, hidden_dim)) for i in range(num_matching)]
```

Comment:
`torch.rand()` isn't a great way to initialize parameters.  Can you use something more reasonable as a default here, like `torch.nn.init.xavier_uniform_`?

Also, having a list of parameters here is pretty hard to interpret if you're looking at parameter values in tensorboard, or during debugging, or similar.  I think it'd be better to give these parameters names, like `full_match_weights`, and assign them directly to `self`, assigning `None` if they aren't selected.  That would also let you get rid of `num_matching`, `mv_idx`, and `mv_idx_increment` entirely, which are a little confusing.

Reply:
Yeah, I know you re-initialize them in your config, but the default should also do something reasonable, in case people forget or don't want to use a special initialization.